File: apps/android/feature/ai/domain/src/main/java/com/scribblefit/feature/ai/domain/usecase/SyncWorkoutUseCase.kt
package com.scribblefit.feature.ai.domain.usecase

import com.scribblefit.feature.ai.domain.engine.ConfigRepository
import com.scribblefit.feature.ai.domain.engine.LLMEngine
import com.scribblefit.feature.ai.domain.engine.SyncRepository
import com.scribblefit.feature.ai.domain.engine.TelemetryRepository
import com.scribblefit.feature.ai.domain.model.ParsingStatus
import com.scribblefit.feature.ai.domain.model.SyncStatus
import com.scribblefit.feature.ai.domain.model.TelemetryData
import kotlinx.coroutines.flow.first

class SyncWorkoutUseCase(
    private val syncRepository: SyncRepository,
    private val telemetryRepository: TelemetryRepository,
    private val engine: LLMEngine,
    private val configRepository: ConfigRepository
) {
    suspend operator fun invoke() {
        val config = configRepository.getConfig().first()
        val promptVersion = config?.promptVersion ?: "1.0.0"

        val pendingItems = syncRepository.getPendingSyncItems().first()

        for (item in pendingItems) {
            syncRepository.updateSyncStatus(item.id, SyncStatus.PROCESSING)

            val result = engine.parseWorkout(item.rawText)

            if (result.status == ParsingStatus.SUCCESS && result.workout != null) {
                syncRepository.saveParsedWorkout(item.id, result.workout)
            } else {
                syncRepository.updateSyncStatus(item.id, SyncStatus.FAILED)

                telemetryRepository.reportError(
                    TelemetryData(
                        rawText = item.rawText,
                        promptVersion = promptVersion,
                        errorMessage = result.error ?: "Unknown error during parsing"
                    )
                )
            }
        }
    }
}

File: apps/ios/ScribbleFit/ScribbleFit/Features/AI/Domain/UseCases/SyncWorkoutUseCase.swift
import Foundation

@MainActor
public final class SyncWorkoutUseCase {
    private let syncRepository: SyncRepository
    private let engine: LLMEngine
    private let configRepository: ConfigRepository
    
    public init(
        syncRepository: SyncRepository,
        engine: LLMEngine,
        configRepository: ConfigRepository
    ) {
        self.syncRepository = syncRepository
        self.engine = engine
        self.configRepository = configRepository
    }
    
    public func execute() async throws {
        let pendingItems = try await syncRepository.getPendingSyncItems()
        
        for item in pendingItems {
            guard let rawText = item.rawText else { continue }
            try await syncRepository.updateSyncStatus(id: item.id, status: .processing)
            
            let result = await engine.parseWorkout(rawText: rawText)
            
            if result.status == .success, let workout = result.workout {
                try await syncRepository.saveParsedWorkout(syncItemId: item.id, workout: workout)
            } else {
                try await syncRepository.updateSyncStatus(id: item.id, status: .failed)
            }
        }
    }
}
